modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/serializer/RecordDataV1Serializer.java

```diff
@@ -287,57 +278,45 @@ private boolean needEncryption(int grpId) {
      * @throws IOException If failed.
      * @throws IgniteCheckedException If failed.
      */
-    private T3<ByteBufferBackedDataInput, Integer, RecordType> readEncryptedData(
+    private EncryptedData readEncryptedData(
         ByteBufferBackedDataInput in,
         boolean readType,
         boolean readKeyId
     ) throws IOException, IgniteCheckedException {
         int grpId = in.readInt();
         int encRecSz = in.readInt();
 
-        RecordType plainRecType = null;
-
-        if (readType)
-            plainRecType = RecordV1Serializer.readRecordType(in);
+        RecordType plainRecType = readType ? RecordV1Serializer.readRecordType(in) : null;
 
         int keyId = readKeyId ? in.readUnsignedByte() : GridEncryptionManager.INITIAL_KEY_ID;
 
-        byte[] encData = new byte[encRecSz];
+        // Encryption Manager can be null during offline WAL iteration
+        if (encMgr == null) {
+            int skipped = in.skipBytes(encRecSz);
 
-        in.readFully(encData);
+            assert skipped == encRecSz;
 
-        GroupKey grpKey = encMgr.groupKey(grpId, keyId);
+            return new EncryptedData(null, plainRecType, grpId);
+        }
 
-        if (grpKey == null)
-            return new T3<>(null, grpId, plainRecType);
+        GroupKey grpKey = encMgr.groupKey(grpId, keyId);
 
-        byte[] clData = encSpi.decrypt(encData, grpKey.key());
+        // Encryption key is not available when restoring the MetaStorage
+        if (grpKey == null) {
+            int skipped = in.skipBytes(encRecSz);
```

Comment:
This block duplicates the block above. How about extracting it to a tiny method?

Reply:
This block is three lines of code and used in only two places. I think it is better as it is

Reply:
Ok, it's not a big deal